vllm-ext/vllm/extension/ns/__init__.py

```diff
@@ -0,0 +1,174 @@
+# ==========================================================================================
+_IE_MODEL: IE_Model = None
+
+# global request_id to internal
+
+
+# allow ns quantization
+# ==========================================================================================
+vllm_config = importlib.import_module("vllm.config")
+old_verify_quantization = vllm_config.ModelConfig._verify_quantization
+
+
+def _verify_quntization(self):
+    if self.quantization is not None and self.quantization == "ns":
+        os.environ["NS_QUANTIZATION"] = "1"
+        return
+    return old_verify_quantization(self)
+
+
+vllm_config.ModelConfig._verify_quantization = _verify_quntization
+
+
+quant = importlib.import_module("vllm.model_executor.layers.quantization")
+
+from vllm.extension.ns.quantization.cpu_ns_config import NSQuantConfig
+
+quant.QUANTIZATION_METHODS["ns"] = NSQuantConfig
+
+logger.info("__ns extension: add ns quantization config, %s", NSQuantConfig.__name__)
+
+# use ns model loader for ns
+# ==========================================================================================
+vllm_loader = importlib.import_module("vllm.model_executor.model_loader.loader")
+old_get_model_loader = vllm_loader.get_model_loader
+
+from vllm.extension.ns.model.ns_loader import NSModelLoaderV2
+from vllm.extension.ns.quantization.cpu_ns_config import copy_quant_ns_config
+from vllm.config import LoadConfig, ModelConfig
+
+
+def get_model_loader_ns(load_config: LoadConfig) -> vllm_loader.BaseModelLoader:
+    if os.environ["NS_QUANTIZATION"] == "1":
+        return NSModelLoaderV2(load_config)
+    return old_get_model_loader(load_config)
+
+
+vllm_loader.get_model_loader = get_model_loader_ns
+
+
+def _get_linear_method_ns(
+    model_config: vllm_config.ModelConfig, load_config: LoadConfig
+) -> Optional["LinearMethodBase"]:
+    """Get the (maybe quantized) linear method."""
+    linear_method = None
+    if model_config.quantization is not None:
+        quant_config = get_quant_config(model_config, load_config)
+        linear_method = quant_config.get_linear_method()
+    return linear_method
+
+
+vllm_loader._get_linear_method = _get_linear_method_ns
+
+
+# bypass cuda capability check since vllm+cpu doesn't have cuda installed
+def _get_quantization_config(
+    model_config: ModelConfig, load_config: LoadConfig
+) -> Optional[QuantizationConfig]:
+    """Get the quantization config."""
+    if model_config.quantization is not None:
+        # copy quant_ns_config.json to downloaded model directory
+        copy_quant_ns_config(model_config, load_config)
+        return get_quant_config(model_config, load_config)
+    return None
+
+
+vllm_loader._get_quantization_config = _get_quantization_config
+
+# reload to make above changes take effect
+model_loader = importlib.import_module("vllm.model_executor.model_loader")
+importlib.reload(model_loader)
+
+logger.info("__ns extension: use ns model loader for ns model, %s", NSModelLoaderV2.__name__)
```

Comment:
could you summerize a list of which properties are monkey-patched with NS classes in the PR descriptions ?

Reply:
All monkey-patches are in this __init__.py. They are,

__init__.py:52] __ns extension: add ns quantization config, NSQuantConfig
__init__.py:105] __ns extension: use ns model loader for ns model, NSModelLoaderV2
__init__.py:116] __ns extension: replace LlamaModel with ns LLamaModel, NSLLamaModel
__init__.py:136] __ns extension: use ns cache engine for ns, NSCPUCacheEngine
__init__.py:146] __ns extension: replace execute_model in cpu_model_runner, execute_model
__init__.py:171] __ns extension: replace BlockSpaceManager.get_block_space_manager_class in vllm.core.interfaces with get_block_space_manager_class